File: 013_MaxNumberOfKSumPairs.py
"""
Topic: Hashmap / Two Sum Variant
Max Number of K-Sum Pairs: greedy with counts
Time Complexity: O(n)
Space Complexity: O(n)
"""

from collections import Counter

def max_operations(nums, k):
    cnt = Counter(nums)
    res = 0
    for x in list(cnt.keys()):
        while cnt[x] > 0:
            need = k - x
            if need not in cnt or cnt[need] == 0:
                break
            if need == x:
                pairs = cnt[x] // 2
                res += pairs
                cnt[x] -= pairs * 2
                break
            else:
                use = min(cnt[x], cnt[need])
                res += use
                cnt[x] -= use
                cnt[need] -= use
                break
    return res


if __name__ == '__main__':
    print(max_operations([1,2,3,4], 5))  # 2 (1+4,2+3)
    print(max_operations([3,1,3,4,3], 6))  # 1 (3+3)

File: 013_MaxNumberOfKSumPairs.cpp
// Topic: Hashmap / Greedy
// Max Number of K-Sum Pairs: Find the maximum number of pairs in an array that sum to k.
// Time Complexity: O(n)
// Space Complexity: O(n)

#include <bits/stdc++.h>
using namespace std;

int maxOperations(vector<int>& nums, int k)
{
    unordered_map<int, int> cnt;
    int ops = 0;
    for (int x : nums)
    {
        int need = k - x;
        auto it = cnt.find(need);
        if (it != cnt.end() && it->second > 0)
        {
            --(it->second);
            ++ops;
        }
        else
        {
            ++cnt[x];
        }
    }
    return ops;
}

int main()
{
    vector<int> a1 = {1, 2, 3, 4};
    cout << maxOperations(a1, 5) << '\n';  // 2
    vector<int> a2 = {3, 1, 3, 4, 3};
    cout << maxOperations(a2, 6) << '\n';  // 1
    vector<int> a3 = {3, 3, 3, 3};
    cout << maxOperations(a3, 6) << '\n';  // 2
    return 0;
}
